Framework/API/src/IndexProperty.cpp

```diff
@@ -0,0 +1,163 @@
+
+bool IndexProperty::isDefault() const {
+  // The default value is an empty vector/string.
+  return value().empty();
+}
+
+std::string IndexProperty::isValid() const {
+  std::string error;
+
+  try {
+    getIndices();
+  } catch (std::runtime_error &e) {
+    error = e.what();
+  } catch (std::out_of_range &) {
+    error = "Indices provided to IndexProperty are out of range.";
+  } catch (std::logic_error &) {
+    error = "Duplicate indices supplied to IndexProperty.";
+  }
+
+  if (!m_validString.empty())
+    error += " " + m_validString;
+
+  return error;
+}
+
+std::string IndexProperty::operator=(const std::string &rhs) {
+  return setValue(rhs);
+}
+
+std::vector<int> &IndexProperty::operator=(const std::vector<int> &rhs) {
+  m_indicesExtracted = false;
+  m_min = m_max = 0;
+  m_validString = "";
+
+  // Instead of just copying vector, determine if we have a pure range.
+  if (rhs.size() > 0) {
+    auto res = std::minmax_element(rhs.cbegin(), rhs.cend());
+    auto minIndex = res.first - rhs.begin();
+    auto maxIndex = res.second - rhs.begin();
+    bool isPureRange =
+        static_cast<size_t>((rhs[maxIndex] - rhs[minIndex] + 1)) == rhs.size();
+
+    if (isPureRange) {
+      m_min = rhs[0];
+      m_max = rhs[rhs.size() - 1];
+      m_value = std::vector<int>();
+    } else
+      m_value = rhs;
+  }
+
+  return m_value;
+}
+
+std::string IndexProperty::setValue(const std::string &value) {
+  m_indicesExtracted = false;
+  m_min = m_max = 0;
+
+  // Check whether or not string value is simply a range
+  if ((std::count(value.cbegin(), value.cend(), ':') == 1 ||
+       std::count(value.cbegin(), value.cend(), '-') == 1) &&
+      std::count(value.cbegin(), value.cend(), ',') == 0) {
+    auto pos = value.find(":");
+    pos = pos == std::string::npos ? value.find("-") : pos;
+    auto min = value.substr(0, pos);
+    auto max = value.substr(pos + 1, value.size() - 1);
+
+    std::string err = "";
+
+    try {
+      m_min = stoi(min);
+      m_max = stoi(max);
+    } catch (std::invalid_argument &) {
+      err += value + " could not be converted to integer list.";
+    }
+
+    if (!err.empty())
+      return m_validString = err;
+
+    m_value = std::vector<int>();
+
+    if (m_min >= m_max)
+      err += " min should be less than max";
+    if (m_min < 0 || m_max < 0)
+      err += " min and max values should be greater than zero";
+
+    return m_validString = err;
+  } else {
+    return m_validString = ArrayProperty<int>::setValue(value);
+  }
+}
+
+std::string IndexProperty::value() const {
+  if (m_min != m_max)
+    return std::to_string(m_min) + ":" + std::to_string(m_max);
+  else
+    return ArrayProperty<int>::value();
+}
+
+Indexing::SpectrumIndexSet IndexProperty::getIndices() const {
+  if (m_indicesExtracted)
+    return m_indices;
+
+  auto wksp = boost::dynamic_pointer_cast<MatrixWorkspace>(
+      m_workspaceProp.getWorkspace());
+
+  if (wksp == nullptr)
```

Comment:
Optional: You could combine the declaration, assignment and null check into one line here.